### PR TITLE
feat(protocol-designer): add well overflow timeline warning

### LINKED ISSUE #6495: PD: Add new case to overflow warnings
## WIP

Copy TBD. Iceboxed 2020/11/11

## Background
Currently PD will warn you if you try to dispense more than the capacity of a well into a well. It does not inform you if you add a volume that is below the well capacity and the total volume now in that well is above well capacity.

## Acceptance Criteria
- [ ] When a user creates a Transfer that will cause the total well volume to exceed well capacity, it should trigger the overflow warning.

### PR DESCRIPTION
**Pull request by IanLondon**
# Overview

Closes #6495

# Changelog

- Add new timeline warning for overflows caused by dispense/blowout/dropTip

# Review requests

- [ ] @howisthisnamenottakenyet copy was not specified in ticket, I winged it, please LMK what you'd like
- [ ] Sanity check: does this seem to give you the intended experience?
- [ ] Code review

# Risk assessment

Medium, PD only

### CODE REVIEW

File: protocol-designer/src/step-generation/__tests__/dispenseUpdateLiquidState.test.js [hunk 1]
```diff
@@ -398,4 +461,54 @@ describe('...8-channel pipette', () => {
...
+      { ...dispenseSingleCh150ToA1Args, pipette: 'p300MultiId' },
+      initialLiquidState
+    )
+
+    // ensure result is added to warnings array
+    expect(dispenseSingleCh150ToA1Args.warnings).toEqual([
+      existingFakeWarning,
+      'fake overflow warning here',
+    ])
+  })
```
Comment (shlokamin) [inline protocol-designer/src/step-generation/__tests__/dispenseUpdateLiquidState.test.js:513]: Should the test case for multi channel be here or in `_getOverflowWarnings.test.js`?

From what I understand, `dispenseUpdateLiquidState` doesn't care about whether it's single or multi, it's always going to call `_getOverflowWarnings`. And if that's the case, I'm not sure what this test case is keeping us safe from here

**Review by shlokamin — COMMENTED** [review]
This was fun to read through, I hadn't really ever read through liquid tracking stuff before! Left a few questions/comments.

File: protocol-designer/src/step-generation/getNextRobotStateAndWarnings/_getOverflowWarnings.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import { getLocationTotalVolume } from '../utils/misc'
+import { wellOverflow } from '../warningCreators'
+
+import type { CommandCreatorWarning, LocationLiquidState } from '../types'
+
+// Given a liquid state after dispense/blowout has happened,
+// return array of errors (or empty array).
+export const _getOverflowWarnings = (
+  liquidState: { [well: string]: LocationLiquidState, ... },
+  wellsForTips: Array<string>,
```
[the remark this answers is not in the document]
Comment (IanLondon) [reply protocol-designer/src/step-generation/getNextRobotStateAndWarnings/_getOverflowWarnings.js:15]: it's a list of wells, but not just any list of wells: there is 1 item for single-channel, and 8 items for multichannel where each well corresponds to a tip

[the remark this answers is not in the document]
Comment (IanLondon) [reply protocol-designer/src/step-generation/getNextRobotStateAndWarnings/_getOverflowWarnings.js:15]: Oh I get your point, within this fn it just a list of wells, and tips don't matter :thinking:

**Comment by IanLondon** [comment]
closing bc we decided to punt on this ticket

**Pull request closed without merging**